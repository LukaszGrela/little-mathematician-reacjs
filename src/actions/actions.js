/*
Copyright 2018 ≈Åukasz 'Severiaan' Grela GrelaDesign

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
import { GAME_ADDITION, GAME_SUBTRACTION, GAME_MULTIPLICATION, GAME_DIVISION } from "../gameTypes";

export const INCREASE_SCORE_OF_GAME = 'INCREASE_SCORE_OF_GAME';



export function increaseScoreOfGame(score, game) {
    return {
        type: INCREASE_SCORE_OF_GAME,
        payload: {
            score,
            game
        }
    }
}



/**
 * Config actions
 */
export const ConfigType = {
    GENERAL: 'config:general',
    ADDITION: 'config' + GAME_ADDITION,
    SUBTRACTION: 'config' + GAME_SUBTRACTION,
    MULTIPLICATION: 'config' + GAME_MULTIPLICATION,
    DIVISION: 'config' + GAME_DIVISION,
};

export const CHANGE_NUMBER_OF_QUESTIONS = 'CHANGE_NUMBER_OF_QUESTIONS';

export function changeNumberOfQuestions(count, type = ConfigType.GENERAL) {
    return {
        type: CHANGE_NUMBER_OF_QUESTIONS,
        payload: { questionCount:count, configType:type }
    };
}
export const CHANGE_RANGE_FROM = 'CHANGE_RANGE_FROM';

export function changeRangeFrom(value, type = ConfigType.GENERAL) {
    return {
        type: CHANGE_RANGE_FROM,
        payload: { value, configType:type }
    };
}
export const CHANGE_RANGE_TO = 'CHANGE_RANGE_TO';

export function changeRangeTo(value, type = ConfigType.GENERAL) {
    return {
        type: CHANGE_RANGE_TO,
        payload: { value, configType:type }
    };
}