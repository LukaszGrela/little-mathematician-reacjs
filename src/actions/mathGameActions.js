/*
Copyright 2018 Łukasz 'Severiaan' Grela GrelaDesign

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/


export const NEW_GAME = 'gd:NEW_GAME';
/**
 * Creates the NEW_GAME action object
 * @param {object} config game configuration object
 * @param {object} config.general general settings
 * @param {number} config.general.questionCount how many questions to ask
 * @param {number} config.general.from lower range of tested numbers
 * @param {number} config.general.to upper range of tested numbers
 */
export const newGame = (config) => ({
    type:NEW_GAME,
    config
});



export const NEXT_QUESTION = 'gd:NEXT_QUESTION';
/** Creates NEXT_QUESTION action object */
export const nextQuestion = () => ({
    type:NEXT_QUESTION
});


export const ANSWER_QUESTION = 'gd:ANSWER_QUESTION';
/**
 * Creates ANSWER_QUESTION action object
 * @param {number} answer user answer
 * @param {number} optionId selected option id
 */
export const answerQuestion = (answer, optionId) => ({
    type: ANSWER_QUESTION,
    answer,
    optionId
})


export const ANSWER_RESULT = 'gd:ANSWER_RESULT';
/**
 * Creates ANSWER_RESULT action object
 * @param {boolean} result User answered correctly or not
 * @param {boolean} finished Default false. Was it the last question that was answered
 */
export const answerResult = (result, finished=false) => ({
    type:ANSWER_RESULT,
    result,
    finished
});
export const REPLAY_GAME = 'gd:REPLAY_GAME';
/** Creates REPLAY_GAME action object */
export const replayGame = () => ({
    type:REPLAY_GAME
});