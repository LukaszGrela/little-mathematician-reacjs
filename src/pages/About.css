.about {
    padding-top: 1em;
    text-align: left;
    width: 300px;
    margin: 0 auto;
}
.about .title {
    font-size: 2em;
    line-height: 3em;
}
.about .acknowledgement {
    color: #757575;
}