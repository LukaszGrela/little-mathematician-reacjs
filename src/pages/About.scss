/*
   Copyright 2018 ≈Åukasz 'Severiaan' Grela GrelaDesign

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

@import '../sass/_variables.scss';
.about {
    padding: 1em;
    text-align: left;
    width: $width-game-column;
    margin: 0 auto;
    .title {
        font-size: 2em;
        line-height: 3em;
    }
    .acknowledgement {
        color: $secondary-text-color;
    }
    .git-source {
        padding: 1rem 0;
        .git-hub-icon {
            vertical-align: bottom;
            width: 1.4rem;
            height: 1.4rem;
            padding-right: 0.4rem;
        }
    }
    
}