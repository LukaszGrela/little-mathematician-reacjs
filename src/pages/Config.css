.config {
    padding-top: 50px;
}
.config label {
    display: block;
    padding-bottom: 8px;
}
.config input,
.config select {
    margin-left: 8px;
}