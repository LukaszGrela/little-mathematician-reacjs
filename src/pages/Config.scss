

.config {
    padding-top: 50px;
    label {
        display: block;
        padding-bottom: 8px;
    }
    input,
    select {
        margin-left: 8px;
    }
}
