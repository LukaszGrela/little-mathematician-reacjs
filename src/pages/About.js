/*
   Copyright 2018 ≈Åukasz 'Severiaan' Grela GrelaDesign

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

import React, { Component } from 'react';

import './About.scss'
import GitHubIcon from '../icons/GitHubIcon';

class About extends Component {
    state = {}
    render() {
        return (
            <div className='about'>
                <div className='title'>Little Mathematician</div>
                <div className='content'>
                    Simple Maths game for children made in ReactJS with aim to explore this framework.
                </div>
                <div className='git-source'>
                    <GitHubIcon/><a href="https://github.com/LukaszGrela/little-mathematician-reacjs" 
                    target='_blank' 
                    rel="noopener noreferrer">LukaszGrela/little-mathematician-reacjs</a>
                </div>
                <div className='acknowledgement'>
                    <ul>
                        <li>Icons in menu (and logo) by Leona Grande from the Noun Project, thenounproject.com/</li>
                        <li>Other icons came from material ui icons, www.materialui.co/</li>
                    </ul>
                </div>
            </div>
        );
    }
}

export default About;
