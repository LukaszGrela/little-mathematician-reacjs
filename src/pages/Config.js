/*
   Copyright 2018 Łukasz 'Severiaan' Grela GrelaDesign

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

import React, { Component } from 'react';

import './Config.css'
import connect from 'react-redux/lib/connect/connect';
import { changeNumberOfQuestions,ConfigType, changeRangeTo, changeRangeFrom } from '../actions/actions';

class Config extends Component {


    constructor(props) {
        super(props);
        this.state = Object.assign({}, props);
    }



    handleChange(field, value) {
        let _val = parseInt(value, 10);
        let gCount = this.props.general.questionCount;
        let gFrom = this.props.general.from;
        let gTo = this.props.general.to;
        const { dispatch } = this.props;
        switch (field) {
            case 'general-qn':
                dispatch(changeNumberOfQuestions(_val, ConfigType.GENERAL));
                break;
            case 'general-rf':
                if (_val >= gTo) {
                    dispatch(changeRangeTo(_val + 10, ConfigType.GENERAL));
                }
                //this.setState({ general: { from: _val, to: gTo, questionCount: gCount } });
                dispatch(changeRangeFrom(_val, ConfigType.GENERAL));
                break;
            case 'general-rt':
                if (_val <= gFrom) {
                    dispatch(changeRangeFrom(_val - 10, ConfigType.GENERAL));
                }
                //this.setState({ general: { to: _val, from: gFrom, questionCount: gCount } });
                dispatch(changeRangeTo(_val, ConfigType.GENERAL));
                break;
            default:
                break;
        }
    }


    render() {
        return (
            <div className='config'>
                <form>
                    <div className='general'>
                        <label>
                            Number of questions:
                            <select value={this.props.general.questionCount}
                                onChange={(e) => this.handleChange('general-qn', e.target.value)}>
                                <option value="10">10</option>
                                <option value="25">25</option>
                                <option value="50">50</option>
                                <option value="100">100</option>
                            </select>
                        </label>
                        <label>
                            Number range from:
                            <select value={this.props.general.from}
                                onChange={(e) => this.handleChange('general-rf', e.target.value)}>
                                <option value="0">0</option>
                                <option value="10">10</option>
                                <option value="20">20</option>
                                <option value="30">30</option>
                                <option value="40">40</option>
                            </select>
                        </label>
                        <label>
                            Number range to:
                            <select value={this.props.general.to}
                                onChange={(e) => this.handleChange('general-rt', e.target.value)}>
                                <option value="10">10</option>
                                <option value="20">20</option>
                                <option value="30">30</option>
                                <option value="40">40</option>
                                <option value="50">50</option>
                            </select>
                        </label>
                    </div>
                    <div className='addition'></div>
                    <div className='subtraction'></div>
                    <div className='multiplication'></div>
                    <div className='division'></div>
                </form>
            </div >
        );
    }
}

const mapStateToProps = (state) => {
    console.log('Config#mapStateToProps', state);
    const config = Object.assign({}, state.config);
    return config;
}

export default connect(mapStateToProps)(Config);