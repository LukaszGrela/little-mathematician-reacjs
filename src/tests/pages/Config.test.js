import React from 'react';
import { Config } from '../../pages/Config';
import { shallow } from 'enzyme';
import { ConfigType } from '../../actions/actions';

let wrapper,
    changeNumberOfQuestions,
    changeRangeFrom,
    changeRangeTo,
    config = {
        general: {
            questionCount: '10',
            from: '0',
            to: '10'
        }
    },
    configModified = {
        general: {
            questionCount: '20',
            from: '20',
            to: '30'
        }
    };

beforeEach(() => {
    changeNumberOfQuestions = jest.fn();
    changeRangeFrom = jest.fn();
    changeRangeTo = jest.fn();
    wrapper = shallow(<Config
        config={config}
        changeNumberOfQuestions={changeNumberOfQuestions}
        changeRangeFrom={changeRangeFrom}
        changeRangeTo={changeRangeTo}
    />);
});

test("Should render Config correctly", () => {
    expect(wrapper).toMatchSnapshot();
});

test("Should render Config correctly with modified data", () => {

    wrapper.setProps({
        config: configModified
    });
    expect(wrapper).toMatchSnapshot();
});

test("Should changeNumberOfQuestions for general conf be called", () => {
    const value = parseInt(configModified.general.questionCount, 10);
    const select = wrapper.find('.general select#general-questions');
    // need to find it
    expect(select.length).toBe(1);
    // need to call our function
    select.simulate('change', {
        target: { value }
    });

    expect(changeNumberOfQuestions).toHaveBeenLastCalledWith(value, ConfigType.GENERAL);

});

test('Should changeRangeFrom for general conf be called, "changeRangeTo" be updated', () => {
    const value = parseInt(configModified.general.from);
    const select = wrapper.find('.general select#general-range-from');
    // need to find it
    expect(select.length).toBe(1);
    // need to call our function
    select.simulate('change', {
        target: { value }
    });
    //
    expect(changeRangeFrom).toHaveBeenLastCalledWith(value, ConfigType.GENERAL);
    // because new value is bigger than the to value, 
    // the 'to' value needs to change too to be 10 more
    expect(changeRangeTo).toHaveBeenLastCalledWith(value + 10, ConfigType.GENERAL);
});


test('Should changeRangeFrom for general conf be called, "changeRangeTo" not called', () => {
    const value = parseInt(configModified.general.from);
    const select = wrapper.find('.general select#general-range-from');
    //
    // wrapper.setProps({
    //     config:{general: {
    //         ...config.general,
    //         to: '40'
    //     }}
    // });
    //
    // need to find it
    expect(select.length).toBe(1);
    // need to call our function
    select.simulate('change', {
        target: { value }
    });
    //
    expect(changeRangeFrom).toHaveBeenLastCalledWith(value, ConfigType.GENERAL);
    // because new value is NOT bigger than the to value, 
    // the 'to' value needs NOT change
    // expect(changeRangeTo).not.toHaveBeenLastCalledWith(value + 10, ConfigType.GENERAL);
    console.warn('Review test: Need to check ho to test that the function was not called at all.');
});

test('Should changeRangeTo for general conf be called, "changeRangeFrom" not called', () => {
    const value = parseInt(configModified.general.to);//'30'
    const select = wrapper.find('.general select#general-range-to');

    // need to find it
    expect(select.length).toBe(1);
    // need to call our function
    select.simulate('change', {
        target: { value }
    });
    //
    expect(changeRangeTo).toHaveBeenLastCalledWith(value, ConfigType.GENERAL);
    // because new value is NOT bigger than the to value, 
    // the 'to' value needs NOT change
    // expect(changeRangeFrom).not.toHaveBeenLastCalledWith(value - 10, ConfigType.GENERAL);
    console.warn('Review test: Need to check ho to test that the function was not called at all.');
});


test('Should changeRangeTo for general conf be called, "changeRangeFrom" be updated', () => {
    const value = parseInt(configModified.general.to);//'30'

    wrapper.setProps({
        config: {
            general: {
                ...config.general,
                from: configModified.general.to
            }
        }
    });
    const select = wrapper.find('.general select#general-range-to');

    // needs to be found
    expect(select.length).toBe(1);
    // call it
    select.simulate('change', {
        target: {
            value
        }
    });
    // needs to be called
    expect(changeRangeTo).toHaveBeenLastCalledWith(value, ConfigType.GENERAL);

    // as the from is the same as to, the changeRangeFrom to be called with 10 less value
    expect(changeRangeFrom).toHaveBeenLastCalledWith(value - 10, ConfigType.GENERAL);

});
