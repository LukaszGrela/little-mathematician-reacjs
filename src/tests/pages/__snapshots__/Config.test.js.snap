// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Config correctly 1`] = `
<div
  className="config"
>
  <form>
    <div
      className="general"
    >
      <label>
        Number of questions:
        <select
          id="general-questions"
          onChange={[Function]}
          value="10"
        >
          <option
            value="10"
          >
            10
          </option>
          <option
            value="25"
          >
            25
          </option>
          <option
            value="50"
          >
            50
          </option>
          <option
            value="100"
          >
            100
          </option>
        </select>
      </label>
      <label>
        Number range from:
        <select
          id="general-range-from"
          onChange={[Function]}
          value="0"
        >
          <option
            value="0"
          >
            0
          </option>
          <option
            value="10"
          >
            10
          </option>
          <option
            value="20"
          >
            20
          </option>
          <option
            value="30"
          >
            30
          </option>
          <option
            value="40"
          >
            40
          </option>
        </select>
      </label>
      <label>
        Number range to:
        <select
          id="general-range-to"
          onChange={[Function]}
          value="10"
        >
          <option
            value="10"
          >
            10
          </option>
          <option
            value="20"
          >
            20
          </option>
          <option
            value="30"
          >
            30
          </option>
          <option
            value="40"
          >
            40
          </option>
          <option
            value="50"
          >
            50
          </option>
        </select>
      </label>
    </div>
    <div
      className="addition"
    />
    <div
      className="subtraction"
    />
    <div
      className="multiplication"
    />
    <div
      className="division"
    />
  </form>
</div>
`;

exports[`Should render Config correctly with modified data 1`] = `
<div
  className="config"
>
  <form>
    <div
      className="general"
    >
      <label>
        Number of questions:
        <select
          id="general-questions"
          onChange={[Function]}
          value="20"
        >
          <option
            value="10"
          >
            10
          </option>
          <option
            value="25"
          >
            25
          </option>
          <option
            value="50"
          >
            50
          </option>
          <option
            value="100"
          >
            100
          </option>
        </select>
      </label>
      <label>
        Number range from:
        <select
          id="general-range-from"
          onChange={[Function]}
          value="20"
        >
          <option
            value="0"
          >
            0
          </option>
          <option
            value="10"
          >
            10
          </option>
          <option
            value="20"
          >
            20
          </option>
          <option
            value="30"
          >
            30
          </option>
          <option
            value="40"
          >
            40
          </option>
        </select>
      </label>
      <label>
        Number range to:
        <select
          id="general-range-to"
          onChange={[Function]}
          value="30"
        >
          <option
            value="10"
          >
            10
          </option>
          <option
            value="20"
          >
            20
          </option>
          <option
            value="30"
          >
            30
          </option>
          <option
            value="40"
          >
            40
          </option>
          <option
            value="50"
          >
            50
          </option>
        </select>
      </label>
    </div>
    <div
      className="addition"
    />
    <div
      className="subtraction"
    />
    <div
      className="multiplication"
    />
    <div
      className="division"
    />
  </form>
</div>
`;
